File: task/src/main/scala/com/goexp/galgame/data/piplline/core/Piplline.java
package com.goexp.galgame.data.piplline.core;

import com.goexp.galgame.data.piplline.exception.RuntimeInterruptedException;
import com.goexp.galgame.data.piplline.handler.HandlerConfig;
import com.goexp.galgame.data.piplline.handler.Starter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

public class Piplline {
    final private Logger logger = LoggerFactory.getLogger(Piplline.class);

    final private MessageQueueProxy<Message<?>> msgQueueProxy;


    final private ExecutorService listenerExecutorService = Executors.newSingleThreadExecutor();

    final private Set<HandlerConfig> configs = new HashSet<>();

    final private Starter starter;


    public Piplline(Starter starter) {
        this.starter = starter;

        msgQueueProxy = new MessageQueueProxy<>(1000);
    }

    public Piplline registry(HandlerConfig config) {
        configs.add(config);
        return this;
    }


    private Piplline registry(int mesType, MessageHandler handler, ExecutorService executor) {
        configs.add(new HandlerConfig(mesType, handler, executor));
        return this;
    }

    private Piplline registry(int mesType, MessageHandler handler, int threadCount) {
        configs.add(new HandlerConfig<>(mesType, handler, Executors.newFixedThreadPool(threadCount)));
        return this;
    }


    public Piplline regForCPUType(int mesCode, MessageHandler handler) {
        return registry(mesCode, handler, 2);
    }

    public Piplline regForCPUType(int mesCode, MessageHandler handler, int threadCount) {
        return registry(mesCode, handler, threadCount);
    }

    public Piplline regForCPUType(int mesCode, MessageHandler handler, ExecutorService executor) {
        return registry(mesCode, handler, executor);
    }

    public Piplline regForIOType(int mesCode, MessageHandler handler) {
        return registry(mesCode, handler, 30);
    }

    public Piplline regForIOType(int mesCode, MessageHandler handler, int threadCount) {
        return registry(mesCode, handler, threadCount);
    }

    public Piplline regForIOType(int mesCode, MessageHandler handler, ExecutorService executor) {
        return registry(mesCode, handler, executor);
    }

    public void start() {
        starter.setQueue(msgQueueProxy);

        var mesTypeMap = configs.stream().peek(c -> c.handler().setQueue(msgQueueProxy))
                .collect(Collectors.groupingBy(HandlerConfig::mesCode));

        listenerExecutorService.execute(() -> {

            var running = true;

            while (running) {
                try {
                    final var mes = msgQueueProxy.poll(5, TimeUnit.MINUTES);

                    if (mes != null) {
                        final var configs = mesTypeMap.get(mes.code);

                        if (configs != null)
                            for (var c : configs) {
                                c.executor().execute(() -> {
                                    try {
                                        c.handler().process(mes);
                                    } catch (Exception e) {
                                        e.printStackTrace();
                                    }
                                });
                            }

                    } else {
                        logger.info("listener task time out!!!");
                        running = false;

                        for (var config : configs) {
                            config.executor().shutdown();
                        }
                        listenerExecutorService.shutdown();
                    }
                } catch (RuntimeInterruptedException e) {
                    e.printStackTrace();
                    running = false;
                }


            }
        });

        starter.process();

    }
}

File: task/src/main/scala/com/goexp/galgame/data/piplline/core/Piplline.scala
package com.goexp.galgame.data.piplline.core

import java.util.concurrent.{ExecutorService, Executors, TimeUnit}

import com.goexp.galgame.data.piplline.exception.RuntimeInterruptedException
import com.goexp.galgame.data.piplline.handler.HandlerConfig
import org.slf4j.LoggerFactory

import scala.collection.mutable

class Piplline(private[this] val starter: Starter) {


  private val logger = LoggerFactory.getLogger(classOf[Piplline])

  private var msgQueueProxy = new MessageQueueProxy[Message](1000)


  private val listenerExecutorService = Executors.newSingleThreadExecutor

  private val configs = mutable.Set[HandlerConfig]()


  def registry(config: HandlerConfig): Piplline = {
    configs.add(config)
    return this
  }


  private def registry(mesType: Int, handler: MessageHandler, executor: ExecutorService): Piplline = {
    configs.add(HandlerConfig(mesType, handler, executor))
    return this
  }

  private def registry(mesType: Int, handler: MessageHandler, threadCount: Int): Piplline = {
    configs.add(new HandlerConfig(mesType, handler, Executors.newFixedThreadPool(threadCount)))
    return this
  }


  def regForCPUType(mesCode: Int, handler: MessageHandler): Piplline = {
    return registry(mesCode, handler, Runtime.getRuntime.availableProcessors)
  }

  def regForCPUType(mesCode: Int, handler: MessageHandler, threadCount: Int): Piplline = {
    return registry(mesCode, handler, threadCount)
  }

  def regForCPUType(mesCode: Int, handler: MessageHandler, executor: ExecutorService): Piplline = {
    return registry(mesCode, handler, executor)
  }

  def regForIOType(mesCode: Int, handler: MessageHandler): Piplline = {
    return registry(mesCode, handler, 30)
  }

  def regForIOType(mesCode: Int, handler: MessageHandler, threadCount: Int): Piplline = {
    return registry(mesCode, handler, threadCount)
  }

  def regForIOType(mesCode: Int, handler: MessageHandler, executor: ExecutorService): Piplline = {
    return registry(mesCode, handler, executor)
  }


  def regGroup(group: Set[HandlerConfig]): Piplline = {
    configs.addAll(group)
    this
  }

  def start(): Unit = { //fill queue
    starter.setQueue(msgQueueProxy)
    val mesTypeMap = configs.to(LazyList)
      .map(c => {
        c.handler.setQueue(msgQueueProxy)
        c
      })
      .groupBy(_.mesCode)
    //      .collect(Collectors.groupingBy(HandlerConfig.mesCode))
    //start message driven
    listenerExecutorService.execute(() => {
      var running = true
      while ( {
        running
      }) try {
        val mes = msgQueueProxy.poll(5, TimeUnit.MINUTES)
        if (mes != null) {
          mesTypeMap.get(mes.code) match {
            case Some(configs) =>
              for (c <- configs) {
                c.executor.execute(() => {
                  try c.handler.process(mes)
                  catch {
                    case e: Exception =>
                      e.printStackTrace()
                  }

                })
              }
            case _ =>
          }
        }
        else {
          logger.info("listener task time out!!!")
          running = false
          for (config <- configs) {
            config.executor.shutdown()
          }
          listenerExecutorService.shutdown()
        }
      } catch {
        case e: RuntimeInterruptedException =>
          e.printStackTrace()
          running = false
      }
    })


    starter.process()
  }
}
